library/nametab.ml

```diff
@@ -378,6 +368,30 @@ module UnivIdMap = HMap.Make(UnivIdOrdered)
 type univrevtab = full_path UnivIdMap.t
 let the_univrevtab = Summary.ref ~name:"univrevtab" (UnivIdMap.empty : univrevtab)
 
+(* Module-related nametab *)
+
+type modules_nametab = {
+  modtypetab : MPTab.t;
+  modtab : MPDTab.t;
+  dirtab : DirTab.t;
+  modrevtab : mprevtab;
+  modtyperevtab : mptrevtab;
+}
+
+let initial_modules_nametab = {
+  modtypetab = MPTab.empty;
+  modtab = MPDTab.empty;
+  dirtab = DirTab.empty;
+  modrevtab = MPmap.empty;
+  modtyperevtab = MPmap.empty
+}
+
+let modules_nametab, modules_nametab_summary_tag =
+  Summary.ref_tag ~name:"MODULES-NAMETAB" initial_modules_nametab
+
+let freeze_modules_nametab () = !modules_nametab
+let unfreeze_modules_nametab = (:=) modules_nametab
```

Comment:
Just a minor nitpick in case you need to modify the pull request before it is merged. Never use the idiom above, as it confuses OCaml to no end. It causes it to allocate a closure (and then to resolve it when called). Always eta-expand module values, i.e., `let unfreeze_modules_nametab x = modules_nametab := x`.

(Perhaps Flambda knows how to compile this idiom correctly, but since we are not using Flambda, this does not matter.)

Reply:
I agree that all functions should always be eta-expanded in OCaml. It's easy to introduce bugs with partial applications. (Here it's innocuous, but see e.g. #15726 for a nasty example.)

Reply:
Sure, I can fix it. Grepping the codebase seems to show quite a few occurrences, btw.

Reply:
Fixed.

Reply:
@maximedenes Can you paste your grep results?

Reply:
Ok, but that's for a (potential) separate patch:

```
lib/flags.ml:    let () = List.iter2 (:=) ol vl in r
lib/flags.ml:    let () = List.iter2 (:=) ol vl in
lib/flags.ml:let set_inline_level = (:=) inline_level
library/summary.ml:      unfreeze_function = ((:=) r);
plugins/extraction/common.ml:  let ph = ref Impl in ((:=) ph), (fun () -> !ph)
plugins/extraction/common.ml:  ((:=) k), (fun () -> !k)
plugins/ltac/tauto.ml:      optwrite = (:=) negation_unfolding }
pretyping/detyping.ml:      optwrite = (:=) print_evar_arguments }
pretyping/program.ml:let set_proofs_transparency = (:=) proofs_transparency
pretyping/program.ml:    optwrite = (:=) program_cases }
pretyping/program.ml:    optwrite = (:=) program_generalized_coercion }
printing/ppconstr.ml:  let set_term_pr = (:=) term_pr
tactics/auto.ml:      optwrite = (:=) refe })
tactics/class_tactics.ml:  let set_typeclasses_debug d = (:=) typeclasses_debug (if d then 1 else 0)
tactics/eauto.ml:      optwrite = (:=) global_debug_eauto })
tactics/eauto.ml:      optwrite = (:=) global_info_eauto })
vernac/attributes.ml:        optwrite = ((:=) b) }
vernac/comSearch.ml:      optwrite = (:=) search_output_name_only }
vernac/prettyp.ml:let set_object_pr = (:=) object_pr
vernac/printmod.ml:      optwrite = ((:=) short) }
vernac/proof_using.ml:      optwrite = ((:=) suggest_proof_using) })
vernac/vernacentries.ml:      optwrite = ((:=) Flags.quiet) }
vernac/vernacentries.ml:      optwrite = (:=) Vmbytegen.dump_bytecode }
vernac/vernacentries.ml:      optwrite = (:=) Vmlambda.dump_lambda }
vernac/vernacinterp.ml:      optwrite = ((:=) default_timeout) }
```